src/v2/catalog.rs

```diff
@@ -1,19 +1,19 @@
 use crate::errors::{Error, Result};
 use crate::v2;
-use futures::{self, stream, Future, Stream};
-use reqwest::StatusCode;
-use serde_json;
+use futures::{self, stream::{self, BoxStream, StreamExt}};
+use reqwest::{RequestBuilder, StatusCode};
+use std::pin::Pin;
 
 /// Convenience alias for a stream of `String` repos.
-pub type StreamCatalog = Box<dyn futures::Stream<Item = String, Error = Error>>;
+pub type StreamCatalog<'a> = BoxStream<'a, Result<String>>;
 
 #[derive(Debug, Default, Deserialize, Serialize)]
 struct Catalog {
     pub repositories: Vec<String>,
 }
 
 impl v2::Client {
-    pub fn get_catalog(&self, paginate: Option<u32>) -> StreamCatalog {
+    pub fn get_catalog<'a, 'b: 'a>(&'b self, paginate: Option<u32>) -> StreamCatalog<'a> {
```

Comment:
When I started to take at look at porting dkregistry to async/.await, I was pointed towards https://github.com/tokio-rs/async-stream for stream handling. Do you by chance have experience with it, or are otherwise willing to compare it to what's currently in this PR?

Reply:
I don't have any experience with it. Do you mind to clean this up in another PR?

Reply:
That's fine by me.

Reply:
After taking another look at the code and spotting the `unsafe` which is used within this function, I suggest the following:

```suggestion
    pub async fn get_catalog(&self, paginate: Option<u32>) -> Result<String> {
```

Reply:
`Result<String>` seems a bit limited too me because as a user I would have to parse the string. I think the method should return `Result<Vec<String>>`. What do you think?

Reply:
`Result<String>` matches the previous ` Box<dyn futures::Stream<Item = String, Error = Error>>`. If we return a `Vec<String>` here, the caller will collect into a `Vec<Vec<String>>` which then needs to be flattened to `Vec<String>`. It would be more convenient for the consumer to collect into that directly. #140 implements my suggestion.

Reply:
Okay. I wasn't aware that you can use futures for streams. But it makes totally sense.